File: src/parameters.cpp
#include "parameters.h"

bool parameter_set = false;

string map_file;

// command types
int CANCEL_JOB;
int LEFT;
int MIDDLE;
int RIGHT;
int NOTHING;

// topics
string SPEECH_SERVICE;
string LISTENING_SERVICE;
string USR_CMD_TOPIC;
string CTRL_TOPIC;
string FORCE_TOPIC;
string ENCODER_TOPIC;
string RESET_PATH_SERVICE;
string PLANNER_ARROW_TOPIC;
string GLOBAL_MAP_TOPIC;
string PLANNED_PATH_TOPIC;


template<typename T>
T readParam(ros::NodeHandle &n, string name)
{
  T ans;
  if (n.getParam(name, ans))
  {
    ROS_INFO_STREAM("Loaded " << name << ": " << ans);
  } else
  {
    ROS_ERROR_STREAM("Failed to load " << name);
    n.shutdown();
  }
  return ans;
}


void readParameters(ros::NodeHandle &n)
{
    // topics
    SPEECH_SERVICE = readParam<string>(n, "speech_service");
    LISTENING_SERVICE = readParam<string>(n, "listening_service");
    USR_CMD_TOPIC = readParam<string>(n, "usr_cmd_topic");

    CTRL_TOPIC = readParam<string>(n, "ctrl_topic");
    FORCE_TOPIC = readParam<string>(n, "force_topic");
    ENCODER_TOPIC = readParam<string>(n, "encoder_topic");

    RESET_PATH_SERVICE = readParam<string>(n, "reset_path_service");
    PLANNER_ARROW_TOPIC = readParam<string>(n, "planner_arrow_topic");
    GLOBAL_MAP_TOPIC = readParam<string>(n, "global_map_topic");
    PLANNED_PATH_TOPIC = readParam<string>(n, "planned_path_topic");

    // command types
    auto state_map = readParam<XmlRpc::XmlRpcValue>(n, "states_map");
    for (int i = 0; i < state_map.size(); i++) {
        auto sublist = state_map[i];
        std::stringstream ss;
        ss << sublist["name"];
        string name = ss.str();
        int value = sublist["key"];
        ROS_INFO("Parameter name: %s, value, %d", name.c_str(), value);
        if (name == "cancel")
            CANCEL_JOB = value;
        else if (name == "left")
            LEFT = value;
        else if (name == "middle")
            MIDDLE = value;
        else if (name == "right")
            RIGHT = value;
        else if (name == "nothing")
            NOTHING = value;
        else
            ROS_ERROR("Parameter %s in states_map is not needed", name);
    } 

    map_file = readParam<string>(n, "map_file");

    parameter_set = true;  
}

File: scripts/parameters.py
import rospy


class Parameters:
    def __init__(self):
        self.parameter_set = False
        self.keywords_map = []
        self.states_map = dict()

    def initParameters(self):
        self.keywords_map = rospy.get_param('keywords_map')
        states = rospy.get_param('states_map')
        for st in states:
            self.states_map[st['name']] = st['key']

        self.speech_service = rospy.get_param('speech_service')
        self.listening_service = rospy.get_param('listening_service')
        self.usr_cmd_topic = rospy.get_param('usr_cmd_topic')
        self.ctrl_topic = rospy.get_param('ctrl_topic')
        self.force_topic = rospy.get_param('force_topic')
        self.encoder_topic = rospy.get_param('encoder_topic')
        self.voltage_topic = rospy.get_param('voltage_topic')

        self.reset_path_service = rospy.get_param('reset_path_service')
        self.planner_arrow_topic = rospy.get_param('planner_arrow_topic')
        self.global_map_topic = rospy.get_param('global_map_topic')
        self.planned_path_topic = rospy.get_param('planned_path_topic')

        self.lidar_original_topic = rospy.get_param('lidar_original_topic')
        self.lidar_sync_topic = rospy.get_param('lidar_sync_topic')

        self.model_path = rospy.get_param("model_path")
        self.map_file = rospy.get_param("map_file")

        self.manual_control = rospy.get_param("manual_control")
        self.init_x = rospy.get_param("init_x")
        self.init_y = rospy.get_param("init_y")
        self.init_theta = rospy.get_param("init_theta")

        self.debug_odometry = rospy.get_param("debug_odometry")
        self.debug_time = rospy.get_param("debug_time")
        self.debug_linear = rospy.get_param("debug_linear")
        self.debug_angular = rospy.get_param("debug_angular")

        self.use_audio = rospy.get_param("use_audio")

        self.parameter_set = True
